### PR TITLE
refactor(ctrlp): add core agent

### PR DESCRIPTION
**Pull request by tiagolobocastro**
Refactor the node,pool,volume agent into a single core agent.
Tidy up the Errors throughout the control plane stack allowing us to
 get more concise error kinds at the rest layer and also at each agent
 endpoint, which will allow us to expose the correct http responses
 at the openapi spec, which will be done as another PR.

### CODE REVIEW

**Comment by tiagolobocastro** [comment]
bors try

File: control-plane/agents/common/src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+                Code::Aborted => ReplyErrorKind::Aborted,
+                Code::OutOfRange => ReplyErrorKind::OutOfRange,
+                Code::Unimplemented => ReplyErrorKind::Unimplemented,
+                Code::Internal => ReplyErrorKind::Internal,
+                Code::Unavailable => ReplyErrorKind::Unavailable,
+                Code::DataLoss => ReplyErrorKind::Internal,
+                Code::Unauthenticated => ReplyErrorKind::Unauthenticated,
+                Code::__NonExhaustive => ReplyErrorKind::Internal,
+            };
+            let extra = format!("{}::{}", request, source.to_string());
```
Comment (cjones1024) [inline control-plane/agents/common/src/errors.rs:274]: Is it worth implementing a From trait here to convert a Code to a ReplyErrorKind

**Review by cjones1024 — APPROVED** [review]
Looks reasonable

File: control-plane/agents/common/src/errors.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/agents/common/src/errors.rs:274]: Yes I think so :+1:

Comment (tiagolobocastro) [reply control-plane/agents/common/src/errors.rs:274]: Oh right, we can't because both types aren't local to the crate...

**Review by paulyoong — APPROVED** [review]

**Review by gila — APPROVED** [review]

File: control-plane/agents/common/src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+use tonic::Code;
+
+/// Common error type for send/receive
+#[derive(Debug, Snafu)]
+#[snafu(visibility = "pub")]
+#[allow(missing_docs)]
+pub enum SvcError {
+    #[snafu(display("Failed to get node '{}' from the node service", node))]
+    BusGetNode { node: String, source: BusError },
+    #[snafu(display("Failed to get nodes from the node service"))]
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/common/src/errors.rs:19]: Done

File: control-plane/agents/common/src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    #[snafu(display("Node '{}' has invalid gRPC URI '{}'", node_id, uri))]
+    GrpcConnectUri {
+        node_id: String,
+        uri: String,
+        source: http::uri::InvalidUri,
+    },
+    #[snafu(display(
+        "gRPC request '{}' for '{}' failed with '{}'",
+        request,
+        request,
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/common/src/errors.rs:45]: Good spot! Done

File: control-plane/agents/common/src/errors.rs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        Self::NotEnoughResources {
+            source,
+        }
+    }
+}
+
+impl From<SvcError> for ReplyError {
+    fn from(error: SvcError) -> Self {
+        #[allow(deprecated)]
+        let desc: &String = &error.description().to_string();
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/common/src/errors.rs:105]: Sadly it's not a drop-in replacement, to_string prints the whole thing and not the name::type of the enum. I suppose if it gets dropped we'll have to implement it ourselves.

File: control-plane/agents/common/src/errors.rs [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    #[snafu(display(
+        "Not enough suitable pools available, {}/{}",
+        have,
+        need
+    ))]
+    OfPools { have: u64, need: u64 },
+    #[snafu(display("Not enough replicas available, {}/{}", have, need))]
+    OfReplicas { have: u64, need: u64 },
+    #[snafu(display("Not enough nexuses available, {}/{}", have, need))]
+    OfNexuses { have: u64, need: u64 },
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/common/src/errors.rs:299]: Yeah, don't know why I thought `Of` would be a good idea :)

File: control-plane/agents/core/src/core/wrapper.rs [hunk 6]
```diff
@@ -0,0 +1,877 @@
...
+    Nexus,
+    NexusId,
+    RemoveNexusChild,
+    ShareNexus,
+    UnshareNexus,
+};
+use std::{ops::Deref, sync::Arc};
+
+#[async_trait]
+pub trait ClientOps {
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/core/src/core/wrapper.rs:354]: Done

File: control-plane/agents/core/src/core/wrapper.rs [hunk 7]
```diff
@@ -0,0 +1,877 @@
...
+    /// Add replica to list
+    pub fn added_replica(&mut self, replica: &Replica) {
+        self.replicas.push(replica.clone())
+    }
+    /// Remove replica from list
+    pub fn removed_replica(&mut self, uuid: &ReplicaId) {
+        self.replicas.retain(|replica| &replica.uuid != uuid)
+    }
+    /// update replica from list
+    pub fn updated_replica(
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/core/src/core/wrapper.rs:778]: Yep, done.

File: control-plane/agents/core/src/node/service.rs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    v0::msg_translation::RpcToMessageBus,
+};
+use rpc::mayastor::ListBlockDevicesRequest;
+use snafu::{OptionExt, ResultExt};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+/// Node's Service
+#[derive(Debug, Clone)]
+pub(crate) struct Service {
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/core/src/node/service.rs:14]: Not sure, I was thinking of just keeping the internal parts of the core Agent as a service. What do you think?

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**